crates/re_arrow_store/src/store_write.rs

```diff
@@ -259,191 +149,120 @@ impl DataStore {
-                    let table = self
-                        .timeless_components
-                        .entry(self.cluster_key)
-                        .or_insert_with(|| {
-                            PersistentComponentTable::new(self.cluster_key, cell.datatype())
-                        });
-                    table.push_cell(cell)
-                } else {
-                    let table = self
-                        .components
-                        .entry(self.cluster_key)
-                        .or_insert_with(|| ComponentTable::new(self.cluster_key, cell.datatype()));
-                    table.push_cell(&self.config, time_point, cell)
-                };
-
-                Ok(row_idx)
+            index.insert_row(insert_id, generated_cluster_cell, row);
+        } else {
+            for (timeline, time) in timepoint.iter() {
+                let ent_path = ent_path.clone(); // shallow
+                let index = self
+                    .tables
+                    .entry((*timeline, ent_path_hash))
+                    .or_insert_with(|| IndexedTable::new(self.cluster_key, *timeline, ent_path));
+
+                index.insert_row(
+                    &self.config,
+                    insert_id,
+                    *time,
+                    generated_cluster_cell.clone(), /* shallow */
+                    row,
+                );
             }
         }
-    }
-
-    pub fn clear_msg_metadata(&mut self, drop_msg_ids: &ahash::HashSet<MsgId>) {
-        crate::profile_function!();
 
-        self.messages
-            .retain(|msg_id, _| !drop_msg_ids.contains(msg_id));
-    }
-}
+        // This is valuable information even for a timeless timepoint!
+        self.metadata_registry.insert(*row_id, timepoint.clone());
 
-// --- Persistent Indices ---
-
-impl PersistentIndexTable {
-    pub fn new(cluster_key: ComponentName, ent_path: EntityPath) -> Self {
-        Self {
-            cluster_key,
-            ent_path,
-            indices: Default::default(),
-            num_rows: 0,
-            all_components: Default::default(),
-        }
+        Ok(())
     }
 
-    #[allow(clippy::unnecessary_wraps)]
-    pub fn insert(&mut self, row_indices: &IntMap<ComponentName, RowIndex>) -> anyhow::Result<()> {
+    /// Auto-generates an appropriate cluster cell for the specified number of instances and
+    /// transparently handles caching.
+    fn generate_cluster_cell(&mut self, num_instances: u32) -> DataCell {
         crate::profile_function!();
 
-        // 2-way merge, step1: left-to-right
-        //
-        // push new row indices to their associated secondary index
-        for (name, row_idx) in row_indices {
-            let index = self
-                .indices
-                .entry(*name)
-                .or_insert_with(|| vec![None; self.num_rows as usize]);
-            index.push(Some(*row_idx));
-        }
+        if let Some(cell) = self.cluster_cell_cache.get(&num_instances) {
```

Comment:
One optimization we can do here in the future is to use slicing to reuse the same cache for many different sizes. That is: `generate_cluster_cell(42)` could return the `0..42` slice of a very long `UInt64Array`